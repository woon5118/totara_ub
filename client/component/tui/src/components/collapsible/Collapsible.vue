<!--
  This file is part of Totara Enterprise Extensions.

  Copyright (C) 2020 onwards Totara Learning Solutions LTD

  Totara Enterprise Extensions is provided only to Totara
  Learning Solutions LTD's customers and partners, pursuant to
  the terms and conditions of a separate agreement with Totara
  Learning Solutions LTD or its affiliate.

  If you do not have an agreement with Totara Learning Solutions
  LTD, you may not access, use, modify, or distribute this software.
  Please contact [licensing@totaralearning.com] for more information.

  @author Kevin Hottinger <kevin.hottinger@totaralearning.com>
  @module tui
-->

<template>
  <div class="tui-collapsible">
    <div class="tui-collapsible__header">
      <ButtonIcon
        class="tui-collapsible__header_icons"
        :styleclass="{
          transparent: true,
        }"
        :aria-expanded="expanded.toString()"
        :aria-controls="generatedId + 'region'"
        :aria-label="label"
        @click="toggleExpand()"
      >
        <CollapseIcon v-if="expanded" size="100" />
        <ExpandIcon v-else size="100" />
      </ButtonIcon>
      <h3 :id="generatedId + 'label'" class="tui-collapsible__header-text">
        {{ label }}
        <slot name="label-extra" />
      </h3>

      <div class="tui-collapsible__header-sideContent">
        <slot name="collapsible-side-content" />
      </div>
    </div>
    <div
      v-if="alwaysRender || (expanded && $scopedSlots.default)"
      v-show="!alwaysRender || (expanded && $scopedSlots.default)"
      :id="generatedId + 'region'"
      class="tui-collapsible__content"
      role="region"
      :aria-labelledby="generatedId + 'label'"
    >
      <slot :expanded="expanded" />
    </div>
  </div>
</template>

<script>
import ButtonIcon from 'tui/components/buttons/ButtonIcon';
import CollapseIcon from 'tui/components/icons/Collapse';
import ExpandIcon from 'tui/components/icons/Expand';

export default {
  components: {
    ButtonIcon,
    CollapseIcon,
    ExpandIcon,
  },

  props: {
    alwaysRender: Boolean,
    id: {
      type: [String, Number],
    },
    label: {
      required: true,
      type: String,
    },
    initialState: {
      default: false,
      type: Boolean,
    },
    value: {
      default: undefined,
      type: Boolean,
    },
  },

  data() {
    return {
      state: this.initialState,
    };
  },

  computed: {
    /**
     * Update expand state base on value or internal state
     *
     * @return {Bool}
     */
    expanded() {
      // If no value prop provided use internal state
      if (this.value === undefined) {
        return this.state;
      }
      return this.value;
    },

    /**
     * Provide ID for accessibility tags
     *
     * @return {Bool}
     */
    generatedId() {
      return this.id || this.$id();
    },
  },

  methods: {
    // Toggle expanded state
    toggleExpand() {
      // If no value prop provided toggle internal state
      if (this.value === undefined) {
        this.state = !this.state;
        return;
      }
      // Propagate expanded value change to parent
      this.$emit('input', !this.value);
    },
  },
};
</script>

<style lang="scss">
.tui-collapsible {
  &__header {
    display: flex;
    padding: var(--gap-2) var(--gap-2) var(--gap-2) 0;
    background: var(--collapsible-header-bg-color);
    border: var(--border-width-thin) solid
      var(--collapsible-header-border-color);

    &_icons {
      padding-right: var(--gap-4);
      padding-left: var(--gap-4);
    }

    &-text {
      @include tui-font-heading-x-small();
      margin: 0;
    }

    &-sideContent {
      display: flex;
      flex-shrink: 0;
      margin: auto 0 auto auto;
    }
  }

  &__content {
    & > .tui-formRowStack {
      margin-top: var(--gap-4);
    }
  }
}
</style>
