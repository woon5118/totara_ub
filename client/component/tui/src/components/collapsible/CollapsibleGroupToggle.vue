<!--
  This file is part of Totara Enterprise Extensions.

  Copyright (C) 2020 onwards Totara Learning Solutions LTD

  Totara Enterprise Extensions is provided only to Totara
  Learning Solutions LTD's customers and partners, pursuant to
  the terms and conditions of a separate agreement with Totara
  Learning Solutions LTD or its affiliate.

  If you do not have an agreement with Totara Learning Solutions
  LTD, you may not access, use, modify, or distribute this software.
  Please contact [licensing@totaralearning.com] for more information.

  @author Kevin Hottinger <kevin.hottinger@totaralearning.com>
  @module tui
-->

<template>
  <div class="tui-collapsibleGroupToggle">
    <ButtonIcon
      :aria-expanded="allExpanded.toString()"
      :aria-label="$str(allExpanded ? 'collapseall' : 'expandall', 'core')"
      class="tui-collapsibleGroupToggle__button"
      :styleclass="{
        transparent: true,
      }"
      :text="$str(allExpanded ? 'collapseall' : 'expandall', 'core')"
      @click="toggleAllFilters()"
    >
      <CollapseIcon v-if="allExpanded" size="200" />
      <ExpandIcon v-else size="200" />
    </ButtonIcon>
  </div>
</template>

<script>
import ButtonIcon from 'tui/components/buttons/ButtonIcon';
import CollapseIcon from 'tui/components/icons/Collapse';
import ExpandIcon from 'tui/components/icons/Expand';

export default {
  components: {
    ButtonIcon,
    CollapseIcon,
    ExpandIcon,
  },

  props: {
    id: {
      type: [String, Number],
    },
    value: {
      required: true,
      type: Object,
    },
  },

  computed: {
    /**
     * Update expand state base on value
     *
     * @return {Bool}
     */
    allExpanded() {
      if (Object.values(this.value).findIndex(elem => elem === false) >= 0) {
        return false;
      }
      return true;
    },

    /**
     * Provide ID for accessibility tags
     *
     * @return {Bool}
     */
    generatedId() {
      return this.id || this.$id();
    },
  },

  methods: {
    /**
     * Emit updated object
     *
     */
    toggleAllFilters() {
      const newState = !this.allExpanded;
      let stateObj = this.value;

      Object.keys(stateObj).forEach(nestedKey => {
        stateObj[nestedKey] = newState;
      });
      this.$emit('input', stateObj);
    },
  },
};
</script>

<lang-strings>
{
  "core": [
    "expandall",
    "collapseall"
  ]
}
</lang-strings>

<style lang="scss">
.tui-collapsibleGroupToggle {
  display: flex;
  &__button {
    margin-left: auto;
  }
}
</style>
