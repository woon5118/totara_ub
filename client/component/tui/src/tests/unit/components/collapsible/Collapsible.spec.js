/**
 * This file is part of Totara Enterprise Extensions.
 *
 * Copyright (C) 2020 onwards Totara Learning Solutions LTD
 *
 * Totara Enterprise Extensions is provided only to Totara
 * Learning Solutions LTD's customers and partners, pursuant to
 * the terms and conditions of a separate agreement with Totara
 * Learning Solutions LTD or its affiliate.
 *
 * If you do not have an agreement with Totara Learning Solutions
 * LTD, you may not access, use, modify, or distribute this software.
 * Please contact [licensing@totaralearning.com] for more information.
 *
 * @author Simon Chester <simon.chester@totaralearning.com>
 * @module tui
 */

import { shallowMount } from '@vue/test-utils';
import collapsible from 'tui/components/collapsible/Collapsible';
import { axe, toHaveNoViolations } from 'jest-axe';
expect.extend(toHaveNoViolations);
let wrapper;

describe('collapsible', () => {
  beforeAll(() => {
    wrapper = shallowMount(collapsible, {
      propsData: { label: 'hello', value: true, initialState: true },
      mocks: {
        $id: x => 'id' + x,
        $str: function() {
          return 'fff';
        },
      },
      scopedSlots: {
        default({ empty }) {
          return (
            !empty &&
            this.$createElement('div', { class: 'test-action' }, 'hello')
          );
        },
      },
    });
  });

  it('matches snapshot', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('should not have any accessibility violations', async () => {
    const results = await axe(wrapper.element, {
      rules: {
        region: { enabled: false },
      },
    });
    expect(results).toHaveNoViolations();
  });
});
