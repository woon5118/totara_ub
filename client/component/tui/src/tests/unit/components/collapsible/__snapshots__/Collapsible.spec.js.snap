// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`collapsible matches snapshot 1`] = `
<div
  class="tui-collapsible"
>
  <div
    class="tui-collapsible__header"
  >
    <buttonicon-stub
      aria-controls="idundefinedregion"
      ariaexpanded="true"
      arialabel="hello"
      class="tui-collapsible__header_icons"
      styleclass="[object Object]"
      type="button"
    >
      <collapseicon-stub
        size="100"
      />
    </buttonicon-stub>
     
    <h3
      class="tui-collapsible__header-text"
      id="idundefinedlabel"
    >
      
      hello
      
    </h3>
     
    <div
      class="tui-collapsible__header-sideContent"
    />
  </div>
   
  <div
    aria-labelledby="idundefinedlabel"
    class="tui-collapsible__content"
    id="idundefinedregion"
    role="region"
  >
    <div
      class="test-action"
    >
      hello
    </div>
  </div>
</div>
`;
